cmds/dhclient/dhclient.go

```diff
@@ -0,0 +1,186 @@
+package main
+
+import (
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net"
+	"os"
+	"time"
+
+	"github.com/d2g/dhcp4"
+	"github.com/d2g/dhcp4client"
+	"github.com/vishvananda/netlink"
+)
+
+const (
+	defaultIface = "eth0"
+	// slop is the slop in our lease time.
+	slop = 10
+)
+
+var (
+	iList        = []string{defaultIface}
+	renewals     = flag.Int("renewals", 2, "Number of DHCP renewals before exiting")
+	verbose      = flag.Bool("verbose", false, "Verbose output")
+	debug        = func(string, ...interface{}) {}
+	leasetimeout = flag.Int("timeout", 600, "Lease timeout in seconds")
+)
+
+/*
+ * Example Client
+ */
+func dhclient(ifname string, timeout time.Duration, done chan error) {
+	var err error
+
+	// if timeout is < 10 seconds, let's get real.
+	if timeout < slop {
+		timeout = 2 * slop
+	}
+
+	n, err := ioutil.ReadFile(fmt.Sprintf("/sys/class/net/%s/address", ifname))
+	if err != nil {
+		done <- fmt.Errorf("Can't get mac for %v", ifname)
+		return
+	}
+	// This is truly amazing but /sys appends newlines to all this data.
+	n = n[:len(n)-1]
+	m, err := net.ParseMAC(string(n))
+	if err != nil {
+		done <- fmt.Errorf("MAC Error:%v\n", err)
+		return
+	}
+
+	iface, err := netlink.LinkByName(ifname)
+	if err != nil {
+		done <- fmt.Errorf("%s: netlink.LinkByName failed: %v", ifname, err)
+		return
+	}
+
+	c, err := dhcp4client.NewInetSock(dhcp4client.SetLocalAddr(net.UDPAddr{IP: net.IPv4(0, 0, 0, 0), Port: 68}), dhcp4client.SetRemoteAddr(net.UDPAddr{IP: net.IPv4bcast, Port: 67}))
+	if err != nil {
+		done <- fmt.Errorf("Client Conection Generation: %v", err)
+		return
+	}
+
+	client, err := dhcp4client.New(dhcp4client.HardwareAddr(m), dhcp4client.Connection(c), dhcp4client.Timeout(timeout))
+	if err != nil {
+		done <- fmt.Errorf("Error:%v\n", err)
+		return
+	}
+
+	// we require at least one successful request.
+	success, p, err := client.Request()
+
+	if err != nil {
+		networkError, ok := err.(*net.OpError)
+		if ok && networkError.Timeout() {
+			done <- fmt.Errorf("%s: Didn't find a DHCP Server", n)
+			return
+		}
+		done <- fmt.Errorf("%s: Error:%v\n", n, err)
+		return
+	}
+
+	debug("Success on %s:%v\n", n, success)
+	debug("Packet:%v\n", p)
+	debug("Lease is %v seconds\n", p.Secs())
+
+	if !success {
+		done <- fmt.Errorf("We didn't sucessfully get a DHCP Lease?")
+		return
+	} else {
+		log.Printf("IP Received:%v\n", p.YIAddr().String())
+	}
+
+	for i := 0; i < *renewals; i++ {
+		addr := p.YIAddr()
+		// We got here because we got a good packet.
+		o := p.ParseOptions()
+		netmask, ok := o[dhcp4.OptionSubnetMask]
+		if ok {
+			fmt.Printf("OptionSubnetMask is %v\n", netmask)
+		} else {
+			// what do to?
+			netmask = addr
+		}
+		dst := &netlink.Addr{IPNet: &net.IPNet{IP: p.YIAddr(), Mask: netmask}, Label: ""}
+		// Add the address to the iface.
+		if err := netlink.AddrAdd(iface, dst); err != nil {
+			if fmt.Sprintf("%v", err) != "file exists" {
```

Comment:
@rjoleary I'm not sure what you're trying to say, what I linked to doesn't take a nil error.

Usage of os.IsExist:
```go
f, err := os.Open("/foo/where")
if err != nil && os.IsExist(err) {
  return fmt.Errorf("file /foo/where does not exist")
}
```